submariner/multiple-cable-drivers.md

```diff
@@ -0,0 +1,192 @@
+
+There are use cases that involve supporting a mix of cable drivers or cable driver configurations within the same
+cluster set. For example, a cluster set could contain three clusters, two that are on-premise and the other
+in the cloud. Connections to the cloud cluster will need to use encryption while the on-premise clusters can use
+unencrypted connections, as they’re connected by a private link, to avoid the performance overhead of encryption.
+This could be achieved using different cable driver implementations or the same cable driver implementation with
+differing encryption configurations.
+
+## Proposal
+
+This enhancement proposes a mechanism to allow users to specify their intent as to which clusters to connect and how to
+connect them. The default would still be a full mesh with an homogeneous cable driver.
+
+The standard mechanism in Kubernetes to specify intent is by creating a Kubernetes API resource whose effect is eventually
+realized into the cluster's desired state. To extend the Kubernetes API, one could either use the generic Kubernetes
+`ConfigMap` or use a custom resource definition (CRD). There's pros and cons to each but a CRD seems more suitable for
+this enhancement.
+
+The finer-grained connectivity and multiple cable driver use cases could be enabled via the same API or separate ones.
+Having one API may be a bit simpler in some respects but it's not foreseen that both use cases would overlap and they
+could interfere with one another and yield undesired results if the user isn't careful as the semantics differ. Therefore
+two APIs with some similar characteristics will be proposed.
+
+### Finer-grained connectivity API
+
+This following CRD is proposed:
+
+```Go
+type ClusterConnectivityPolicy struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec ClusterConnectivityPolicySpec `json:"spec"`
+}
+
+type ClusterConnectivityPolicySpec struct {
+    // LeftClusterSelector identifies the Cluster resources representing the clusters on one end of a connection. An empty
+    // selector indicates wildcard, ie matches any cluster.
+    LeftClusterSelector metav1.LabelSelector `json:"leftClusterSelector,omitempty"`
+
+    // RightClusterSelector identifies the Cluster resources representing the clusters on the other end of a connection.
+    // An empty selector indicates wildcard, ie matches any cluster.
+    RightClusterSelector metav1.LabelSelector `json:"rightClusterSelector,omitempty"`
+}
+```
+
+The `ClusterConnectivityPolicy` CR defines the criteria for selecting which cluster pairs can connect. To resolve the
+connectivity policy for a remote cluster's `Endpoint`, Submariner would search for a policy whose left and right label
+selectors match the corresponding `Cluster` resources of the local and remote `Endpoints`. If any matching policy is found,
+the clusters are connected, otherwise they are not.
```

Comment:
It's been quite a while since I looked at this - trying to re-familiarize...  If there's no matching policy for a particular remote Endpoint, it won't create a connection. Does that answer your question?

Reply:
Ok so then if you are using the fine grained cluster connectivity it's up to you to make sure that there are no dangling clusters - in other words you need to define enough policies to make sure you avoid the situtation, there are no preset defaults

Reply:
By dangling cluster I assume you mean one that isn't included in a policy. Since the policies are a whitelist then such a cluster is naturally excluded from connectivity. The preset default is full mesh connectivity. Being a whitelist, technically the absence of any policy should mean no connectivity but we can imply a wildcard policy out-of-the-box even if we don't explicitly create it.

The other alternative is to use a blacklist approach but that doesn't seem suitable.

Reply:
yeah that's what I meant by dangling cluster...  what got me thinking about it was if you had 7 clusters and 5 of those matched the policy (you definitely have 5 connected clusters) but the question I had then was regarding the 2 remaining clusters, are they automatically connected...
But I think there might be situations where maybe folks don't want to automatically connect those clusters - so probably better to let the end user create a policy for those situations also...

Reply:
> But I think there might be situations where maybe folks don't want to automatically connect those clusters - so probably better to let the end user create a wild card policy to fall back on...

But then wouldn't they be back to automatic full-mesh and thus defeat the purpose of the whitelist policy?

Reply:
Are we saying that choosing to fine tune the interconnects on some of the clusters precludes you from meshing the rest of the clusters in a cluster set?

I just assumed that the user could add a wild card policy as the lowest priority policy and so, any clusters that didn't match any existing policies would be meshed... (of course they can choose to do something else as the fall back policy)

so it comes down to what the use case is, is it 
1. to let someone fine tune the interconnects for all clusters in a cluster set in one go
2. to let someone fine tune the interconnects for some of the clusters in a cluster set at a time.
3. Both :)

I guess I'm not understanding the scope... :(

Reply:
Referring back to the example use cases outlined above, the purpose is to allow the user to restrict which remote clusters a cluster can access, for security, scalability etc. Connectivity is a binary decision and thus a prioritized policy list with fallback is not suitable, as it is with the cable engine config. 

Let's say a user defines policies A and B. The algorithm would cycle thru the policies - first check policy A; if the clusters match the selectors then connect them; otherwise check B. If neither match then they're not connected. Now if the user added a wildcard policy C then every cluster pair would automatically match and thus be connected. This wouldn't seem desirable.

This is my understanding of the problem domain and how to solve it. Perhaps you have a different idea or something I'm missing.

Reply:
Yeah I'm sure there's various other use cases. it's open for ideas. What would you suggest?

Reply:
I was originally thinking about this from the perspective of selective service import and wondering how we might establish connectivity only where it is needed for the declared service consumption. That's what started me considering it from the perspective of a single cluster. I will explore whether we can define the deny/allow default on a per cluster basis and see if I can come up with something worth proposing.

Reply:
I think you could handle that scenario with a backlist policy. Say the satellite cluster is labeled with "_satellite_" and the other with "_east_", then create a backlist policy with label selectors **satellite** and **!east** - this would disallow _satellite_ connecting with any other cluster except for _east_.. With no whitelist policies defined, it would default to the all-inclusive wild card policy which would allow everything else.